Diagnostic assumes that braced unresolved identifiers are formatting argument

Uplifted from https://github.com/rust-lang/rust/pull/141213#discussion_r2099811526. Given

```rs
struct Foo { x: i32 }

impl Foo {
    fn foo(&self) { _ = {x}; }
}
```

the compiler outputs

```
error[E0425]: cannot find value `x` in this scope
 --> file.rs:4:26
  |
4 |     fn foo(&self) { _ = {x}; }
  |                          ^
  |
  = help: you might have meant to use the available field in a format string: `"{}", self.x`

warning: unnecessary braces around assigned value
 --> file.rs:4:25
  |
4 |     fn foo(&self) { _ = {x}; }
  |                         ^ ^
  |
  = note: `#[warn(unused_braces)]` on by default
help: remove these braces
  |
4 -     fn foo(&self) { _ = {x}; }
4 +     fn foo(&self) { _ = x; }
  |
```

Notice the (unstructured) suggestion ``you might have meant to use the available field in a format string: `"{}", self.x` `` which is obviously incorrect, there's no format string in sight and `fn foo(&self) { _ = "{}", self.x; }` is butchered.

### Meta

```
rustc 1.89.0-nightly (c43786c9b 2025-05-21)
binary: rustc
commit-hash: c43786c9b7b8d8dcc3f9c604e0e3074c16ed69d3
commit-date: 2025-05-21
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.5
```
```

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -26,7 +26,7 @@ use rustc_middle::ty;
 use rustc_session::{Session, lint};
 use rustc_span::edit_distance::{edit_distance, find_best_match_for_name};
 use rustc_span::edition::Edition;
-use rustc_span::{DUMMY_SP, Ident, Span, Symbol, kw, sym};
+use rustc_span::{DUMMY_SP, DesugaringKind, Ident, Span, Symbol, kw, sym};
 use thin_vec::ThinVec;
 use tracing::debug;
 
@@ -980,12 +980,15 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                     AssocSuggestion::Field(field_span) => {
                         if self_is_available {
                             let source_map = self.r.tcx.sess.source_map();
-                            // check if the field is used in a format string, such as `"{x}"`
-                            let field_is_format_named_arg = source_map
+                            let field_is_format_named_arg = matches!(
+                                span.desugaring_kind(),
+                                Some(DesugaringKind::FormatLiteral { .. })
+                            ) && source_map
                                 .span_to_source(span, |s, start, _| {
-                                    Ok(s.get(start - 1..start) == Some("{"))
-                                });
-                            if let Ok(true) = field_is_format_named_arg {
+                                    Ok(s.get(start.saturating_sub(1)..start) == Some("{"))
+                                })
+                                .unwrap_or(false);
+                            if field_is_format_named_arg {
                                 err.help(
                                     format!("you might have meant to use the available field in a format string: `\"{{}}\", self.{}`", segment.ident.name),
                                 );

--- a/tests/ui/resolve/suggestions/sugg-field-in-format-string-issue-141136.rs
+++ b/tests/ui/resolve/suggestions/sugg-field-in-format-string-issue-141136.rs
@@ -9,6 +9,7 @@ impl Foo {
         let _ = format!("{ x}"); //~ ERROR invalid format string: expected `}`, found `x`
         let _ = format!("{}", x); //~ ERROR cannot find value `x` in this scope [E0425]
         println!("{x}"); //~ ERROR cannot find value `x` in this scope [E0425]
+        let _ = {x}; //~ERROR cannot find value `x` in this scope [E0425]
     }
 }
 

--- a/tests/ui/resolve/suggestions/sugg-field-in-format-string-issue-141136.stderr
+++ b/tests/ui/resolve/suggestions/sugg-field-in-format-string-issue-141136.stderr
@@ -14,7 +14,10 @@ error[E0425]: cannot find value `x` in this scope
 LL |         let _ = format!("{x}");
    |                           ^
    |
-   = help: you might have meant to use the available field in a format string: `"{}", self.x`
+help: you might have meant to use the available field
+   |
+LL |         let _ = format!("{self.x}");
+   |                           +++++
 
 error[E0425]: cannot find value `x` in this scope
   --> $DIR/sugg-field-in-format-string-issue-141136.rs:8:27
@@ -22,7 +25,10 @@ error[E0425]: cannot find value `x` in this scope
 LL |         let _ = format!("{x }");
    |                           ^^
    |
-   = help: you might have meant to use the available field in a format string: `"{}", self.x`
+help: you might have meant to use the available field
+   |
+LL |         let _ = format!("{self.x }");
+   |                           +++++
 
 error[E0425]: cannot find value `x` in this scope
   --> $DIR/sugg-field-in-format-string-issue-141136.rs:10:31
@@ -41,8 +47,22 @@ error[E0425]: cannot find value `x` in this scope
 LL |         println!("{x}");
    |                    ^
    |
-   = help: you might have meant to use the available field in a format string: `"{}", self.x`
+help: you might have meant to use the available field
+   |
+LL |         println!("{self.x}");
+   |                    +++++
 
-error: aborting due to 5 previous errors
+error[E0425]: cannot find value `x` in this scope
+  --> $DIR/sugg-field-in-format-string-issue-141136.rs:12:18
+   |
+LL |         let _ = {x};
+   |                  ^
+   |
+help: you might have meant to use the available field
+   |
+LL |         let _ = {self.x};
+   |                  +++++
+
+error: aborting due to 6 previous errors
 
 For more information about this error, try `rustc --explain E0425`.

--- a/tests/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr
+++ b/tests/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr
@@ -34,7 +34,10 @@ error[E0425]: cannot find value `config` in this scope
 LL |         println!("{config}");
    |                    ^^^^^^
    |
-   = help: you might have meant to use the available field in a format string: `"{}", self.config`
+help: you might have meant to use the available field
+   |
+LL |         println!("{self.config}");
+   |                    +++++
 help: a local variable with a similar name exists
    |
 LL -         println!("{config}");

